# Applied Data Science Capstone

This notebook will be use for the Capstone Project of the IBM Applied Data Science Specialization.

Author: RUBEN ADAD

### Import necessary Libraries

In [154]:
import pandas as pd
import numpy as np
import requests
from bs4 import BeautifulSoup
import matplotlib.cm as cm
import matplotlib.colors as colors
import io

## Part 1

### Scrape the Wikipedia page

In [155]:
website_url = requests.get("https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M").text

### Convert to BeautifulSoup object

In [156]:
soup = BeautifulSoup(website_url,'lxml')
print(soup.prettify())

<!DOCTYPE html>
<html class="client-nojs" dir="ltr" lang="en">
 <head>
  <meta charset="utf-8"/>
  <title>
   List of postal codes of Canada: M - Wikipedia
  </title>
  <script>
   document.documentElement.className = document.documentElement.className.replace( /(^|\s)client-nojs(\s|$)/, "$1client-js$2" );
  </script>
  <script>
   (window.RLQ=window.RLQ||[]).push(function(){mw.config.set({"wgCanonicalNamespace":"","wgCanonicalSpecialPageName":false,"wgNamespaceNumber":0,"wgPageName":"List_of_postal_codes_of_Canada:_M","wgTitle":"List of postal codes of Canada: M","wgCurRevisionId":890001695,"wgRevisionId":890001695,"wgArticleId":539066,"wgIsArticle":true,"wgIsRedirect":false,"wgAction":"view","wgUserName":null,"wgUserGroups":["*"],"wgCategories":["Communications in Ontario","Postal codes in Canada","Toronto","Ontario-related lists"],"wgBreakFrames":false,"wgPageContentLanguage":"en","wgPageContentModel":"wikitext","wgSeparatorTransformTable":["",""],"wgDigitTransformTable":["",""],"wg

### Extract HTML table

In [157]:
postal_table = soup.find('table',{'class':'wikitable sortable'})
postal_table

<table class="wikitable sortable">
<tbody><tr>
<th>Postcode</th>
<th>Borough</th>
<th>Neighbourhood
</th></tr>
<tr>
<td>M1A</td>
<td>Not assigned</td>
<td>Not assigned
</td></tr>
<tr>
<td>M2A</td>
<td>Not assigned</td>
<td>Not assigned
</td></tr>
<tr>
<td>M3A</td>
<td><a href="/wiki/North_York" title="North York">North York</a></td>
<td><a href="/wiki/Parkwoods" title="Parkwoods">Parkwoods</a>
</td></tr>
<tr>
<td>M4A</td>
<td><a href="/wiki/North_York" title="North York">North York</a></td>
<td><a href="/wiki/Victoria_Village" title="Victoria Village">Victoria Village</a>
</td></tr>
<tr>
<td>M5A</td>
<td><a href="/wiki/Downtown_Toronto" title="Downtown Toronto">Downtown Toronto</a></td>
<td><a href="/wiki/Harbourfront_(Toronto)" title="Harbourfront (Toronto)">Harbourfront</a>
</td></tr>
<tr>
<td>M5A</td>
<td><a href="/wiki/Downtown_Toronto" title="Downtown Toronto">Downtown Toronto</a></td>
<td><a href="/wiki/Regent_Park" title="Regent Park">Regent Park</a>
</td></tr>
<tr>
<td>M6A</td>

### Extract headings

In [158]:
ths = postal_table.find_all('th')
headings = [th.text.strip() for th in ths]
        
print(headings)
postal_df = pd.DataFrame(columns=headings)

['Postcode', 'Borough', 'Neighbourhood']


### Extract table content

In [159]:
tds = postal_table.find_all('td')
values = [td.text.strip() for td in tds]
        
print(values)
print(type(values))
print(len(values))

['M1A', 'Not assigned', 'Not assigned', 'M2A', 'Not assigned', 'Not assigned', 'M3A', 'North York', 'Parkwoods', 'M4A', 'North York', 'Victoria Village', 'M5A', 'Downtown Toronto', 'Harbourfront', 'M5A', 'Downtown Toronto', 'Regent Park', 'M6A', 'North York', 'Lawrence Heights', 'M6A', 'North York', 'Lawrence Manor', 'M7A', "Queen's Park", 'Not assigned', 'M8A', 'Not assigned', 'Not assigned', 'M9A', 'Etobicoke', 'Islington Avenue', 'M1B', 'Scarborough', 'Rouge', 'M1B', 'Scarborough', 'Malvern', 'M2B', 'Not assigned', 'Not assigned', 'M3B', 'North York', 'Don Mills North', 'M4B', 'East York', 'Woodbine Gardens', 'M4B', 'East York', 'Parkview Hill', 'M5B', 'Downtown Toronto', 'Ryerson', 'M5B', 'Downtown Toronto', 'Garden District', 'M6B', 'North York', 'Glencairn', 'M7B', 'Not assigned', 'Not assigned', 'M8B', 'Not assigned', 'Not assigned', 'M9B', 'Etobicoke', 'Cloverdale', 'M9B', 'Etobicoke', 'Islington', 'M9B', 'Etobicoke', 'Martin Grove', 'M9B', 'Etobicoke', 'Princess Gardens', 'M9B

### Convert to Dataframe

During conversion when Neighbourhood is "not assigned" it will be assigned with Borough name.

Resulting dataframe is sorted by Postcode column and has 288 rows and 3 columns.

In [160]:
j = 0
for i in range(0,len(values),3):
    Postcode = values[i]
    Borough = values[i+1]
    Neighbourhood = values[i+2]

    if Neighbourhood == "Not assigned":
        Neighbourhood = Borough


    postal_df.loc[j] = [Postcode,Borough,Neighbourhood]
    j += 1
    
postal_df.sort_values(by=['Postcode'], inplace=True) 
postal_df.reset_index(inplace=True) 
postal_df.drop('index', axis=1, inplace=True)
print(postal_df.shape)
postal_df.head(25)

(288, 3)


Postcode       Borough             Neighbourhood
0       M1A  Not assigned              Not assigned
1       M1B   Scarborough                   Malvern
2       M1B   Scarborough                     Rouge
3       M1C   Scarborough            Highland Creek
4       M1C   Scarborough                Rouge Hill
5       M1C   Scarborough                Port Union
6       M1E   Scarborough                 West Hill
7       M1E   Scarborough               Morningside
8       M1E   Scarborough                 Guildwood
9       M1G   Scarborough                    Woburn
10      M1H   Scarborough                 Cedarbrae
11      M1J   Scarborough       Scarborough Village
12      M1K   Scarborough              Kennedy Park
13      M1K   Scarborough                   Ionview
14      M1K   Scarborough      East Birchmount Park
15      M1L   Scarborough                  Oakridge
16      M1L   Scarborough               Golden Mile
17      M1L   Scarborough                  Clairlea
18      M1M   Scarborough                Cliffcrest
19      M1M   Scarborough                 Cliffside
20      M1M   Scarborough  Scarborough Village West
21      M1N   Scarborough               Birch Cliff
22      M1N   Scarborough            Cliffside West
23      M1P   Scarborough               Dorset Park
24      M1P   Scarborough   Scarborough Town Centre

### Merge Neighboorhoods for duplicated Postal Codes

In [161]:
# df.set_value(i,'ifor',ifor_val)
last_Postcode = ""
last_Neighbourhood = ""
Neighbourhood = ""

for idx, row in postal_df.iterrows():
    
    if last_Postcode != row['Postcode']:
        j = idx
        Neighbourhood = ""
        #print(last_Postcode, row['Postcode'], j)
    else:
        Neighbourhood = postal_df.at[j, 'Neighbourhood'] + ", " + row['Neighbourhood'] 
        postal_df.at[j, 'Neighbourhood'] = Neighbourhood
        postal_df.drop([idx], inplace=True)
        #print(last_Postcode, row['Postcode'], j, index)

    last_Postcode = row['Postcode']
    last_Neighbourhood = row['Neighbourhood']
    


postal_df.head(15)

Postcode       Borough                                      Neighbourhood
0       M1A  Not assigned                                       Not assigned
1       M1B   Scarborough                                     Malvern, Rouge
3       M1C   Scarborough             Highland Creek, Rouge Hill, Port Union
6       M1E   Scarborough                  West Hill, Morningside, Guildwood
9       M1G   Scarborough                                             Woburn
10      M1H   Scarborough                                          Cedarbrae
11      M1J   Scarborough                                Scarborough Village
12      M1K   Scarborough        Kennedy Park, Ionview, East Birchmount Park
15      M1L   Scarborough                    Oakridge, Golden Mile, Clairlea
18      M1M   Scarborough    Cliffcrest, Cliffside, Scarborough Village West
21      M1N   Scarborough                        Birch Cliff, Cliffside West
23      M1P   Scarborough  Dorset Park, Scarborough Town Centre, Wexford ...
26      M1R   Scarborough                                  Wexford, Maryvale
28      M1S   Scarborough                                          Agincourt
29      M1T   Scarborough            Tam O'Shanter, Clarks Corners, Sullivan

#### Dataframe after merging duplicated Postal Codes has 180 rows.

In [162]:
postal_df.shape

(180, 3)

## Part 2 Geocoding

### Read csv file that has the geographical coordinates of each postal code

In [163]:
url = 'http://cocl.us/Geospatial_data'

urlData = requests.get(url).content
codes_df = pd.read_csv(io.StringIO(urlData.decode('utf-8')))

codes_df.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

### Join geographical coordinates DF with Postal Codes DF

I am using INNER join to avoid rows without geographical coordinates.

Resulting dataframe has 103 rows.

In [164]:
latlon_df = postal_df.set_index('Postcode').join(codes_df.set_index('Postal Code'), how="inner")
latlon_df.reset_index(inplace=True) 
latlon_df.rename(columns={'index': 'Postcode'}, inplace=True)
print(latlon_df.shape)
latlon_df.head()

(103, 5)


Postcode      Borough                           Neighbourhood   Latitude  \
0      M1B  Scarborough                          Malvern, Rouge  43.806686   
1      M1C  Scarborough  Highland Creek, Rouge Hill, Port Union  43.784535   
2      M1E  Scarborough       West Hill, Morningside, Guildwood  43.763573   
3      M1G  Scarborough                                  Woburn  43.770992   
4      M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

### Visualize Postal Codes in a map

In [165]:
import folium
toronto_map = folium.Map(location=[43.653908, -79.384293], zoom_start=10) # generate map centred around Toronto


# add popular spots to the map as blue circle markers
for lat, lng, label in zip(latlon_df['Latitude'], latlon_df['Longitude'], latlon_df['Postcode']):
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        fill=True,
        color='blue',
        fill_color='blue',
        fill_opacity=0.6
        ).add_to(toronto_map)

# display map
toronto_map

## Part 3 Clustering

Import needed libraries:

In [166]:
import json # library to handle JSON files
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values
import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe
from sklearn.cluster import KMeans

Provide Foursqaure API credentials:

In [167]:
CLIENT_ID = 'LZ0PKCF0CXYYTZFP3U4VY0RG4MBURBC0WHZ5TTWHECUMIEPC' # your Foursquare ID
CLIENT_SECRET = 'MWUOBK3YN4Y2LPQSDFG4SMWMGWUKO2J3FZ452OUVECNJDSOU' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: LZ0PKCF0CXYYTZFP3U4VY0RG4MBURBC0WHZ5TTWHECUMIEPC
CLIENT_SECRET:MWUOBK3YN4Y2LPQSDFG4SMWMGWUKO2J3FZ452OUVECNJDSOU


Get Toronto latitude and longitud:

In [168]:
address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print(location, latitude, longitude)

Toronto, Ontario, M6K 1X9, Canada 43.653963 -79.387207


Function that extracts the category of the venue:

In [169]:
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

Function to get information for all the neighborhoods in Toronto:

In [170]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

Execute the getNerbyVenues function:

In [171]:
radius = 500
LIMIT = 100

toronto_venues = getNearbyVenues(names=latlon_df['Postcode'],
                                   latitudes=latlon_df['Latitude'],
                                   longitudes=latlon_df['Longitude']
                                  )

M1B
M1C
M1E
M1G
M1H
M1J
M1K
M1L
M1M
M1N
M1P
M1R
M1S
M1T
M1V
M1W
M1X
M2H
M2J
M2K
M2L
M2M
M2N
M2P
M2R
M3A
M3B
M3C
M3H
M3J
M3K
M3L
M3M
M3N
M4A
M4B
M4C
M4E
M4G
M4H
M4J
M4K
M4L
M4M
M4N
M4P
M4R
M4S
M4T
M4V
M4W
M4X
M4Y
M5A
M5B
M5C
M5E
M5G
M5H
M5J
M5K
M5L
M5M
M5N
M5P
M5R
M5S
M5T
M5V
M5W
M5X
M6A
M6B
M6C
M6E
M6G
M6H
M6J
M6K
M6L
M6M
M6N
M6P
M6R
M6S
M7A
M7R
M7Y
M8V
M8W
M8X
M8Y
M8Z
M9A
M9B
M9C
M9L
M9M
M9N
M9P
M9R
M9V
M9W


Print dataframe shape and first rows:

In [172]:
print(toronto_venues.shape)
toronto_venues.head()

(2238, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0          M1B              43.806686              -79.194353   
1          M1B              43.806686              -79.194353   
2          M1C              43.784535              -79.160497   
3          M1E              43.763573              -79.188711   
4          M1E              43.763573              -79.188711   

                             Venue  Venue Latitude  Venue Longitude  \
0                          Wendy's       43.807448       -79.199056   
1            Interprovincial Group       43.805630       -79.200378   
2            Royal Canadian Legion       43.782533       -79.163085   
3  Swiss Chalet Rotisserie & Grill       43.767697       -79.189914   
4                G & G Electronics       43.765309       -79.191537   

         Venue Category  
0  Fast Food Restaurant  
1            Print Shop  
2                   Bar  
3           Pizza Place  
4     Electronics Store

Keep only neighborhoods with more than 10 venues (new dataframe has 1988 rows compared with 2238 in the original dataframe):

In [173]:
toronto_venues10 = toronto_venues.groupby('Neighborhood').filter(lambda x : len(x)>10)
toronto_venues10.shape

(1988, 7)

Print the number of venues for each neighborhood:

In [174]:
toronto_venues10.groupby('Neighborhood').count()

Neighborhood Latitude  Neighborhood Longitude  Venue  \
Neighborhood                                                         
M1W                              15                      15     15   
M2J                              62                      62     62   
M2N                              35                      35     35   
M3C                              22                      22     22   
M3H                              17                      17     17   
M4B                              13                      13     13   
M4G                              32                      32     32   
M4H                              16                      16     16   
M4K                              44                      44     44   
M4L                              19                      19     19   
M4M                              37                      37     37   
M4R                              18                      18     18   
M4S                              39                      39     39   
M4V                              14                      14     14   
M4X                              42                      42     42   
M4Y                              89                      89     89   
M5A                              46                      46     46   
M5B                             100                     100    100   
M5C                             100                     100    100   
M5E                              56                      56     56   
M5G                              85                      85     85   
M5H                             100                     100    100   
M5J                             100                     100    100   
M5K                             100                     100    100   
M5L                             100                     100    100   
M5M                              26                      26     26   
M5R                              24                      24     24   
M5S                              34                      34     34   
M5T                             100                     100    100   
M5V                              14                      14     14   
M5W                              95                      95     95   
M5X                             100                     100    100   
M6A                              14                      14     14   
M6G                              15                      15     15   
M6H                              19                      19     19   
M6J                              64                      64     64   
M6K                              21                      21     21   
M6P                              25                      25     25   
M6R                              16                      16     16   
M6S                              38                      38     38   
M7A                              39                      39     39   
M7Y                              18                      18     18   
M8V                              13                      13     13   
M8Z                              12                      12     12   

              Venue Latitude  Venue Longitude  Venue Category  
Neighborhood                                                   
M1W                       15               15              15  
M2J                       62               62              62  
M2N                       35               35              35  
M3C                       22               22              22  
M3H                       17               17              17  
M4B                       13               13              13  
M4G                       32               32              32  
M4H                       16               16              16  
M4K                       44               44              44  
M4L                       19               19              19  
M4M                      

In [175]:
print('There are {} uniques categories.'.format(len(toronto_venues10['Venue Category'].unique())))

There are 254 uniques categories.


Dummy coding of "venue category" attribute. We get 1 column for each of the 254 categories. We add the "neighborhood" column and move it as the first column of the dataframe:

In [176]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues10[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Nborhood'] = toronto_venues10['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Nborhood  Accessories Store  Adult Boutique  Afghan Restaurant  Airport  \
71      M1W                  0               0                  0        0   
72      M1W                  0               0                  0        0   
73      M1W                  0               0                  0        0   
74      M1W                  0               0                  0        0   
75      M1W                  0               0                  0        0   

    Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
71                   0             0               0                0   
72                   0             0               0                0   
73                   0             0               0                0   
74                   0             0               0                0   
75                   0             0               0                0   

    Airport Terminal     ...       Train Station  \
71                 0     ...                   0   
72                 0     ...                   0   
73                 0     ...                   0   
74                 0     ...                   0   
75                 0     ...                   0   

    Vegetarian / Vegan Restaurant  Video Game Store  Video Store  \
71                              0                 0            0   
72                              0                 0            0   
73                              0                 0            0   
74                              0                 0            0   
75                              0                 0            0   

    Vietnamese Restaurant  Warehouse Store  Wine Bar  Wings Joint  \
71                      0                0         0            0   
72                      0                0         0            0   
73                      0                0         0            0   
74                      0                0         0            0   
75                      0                0         0            0   

    Women's Store  Yoga Studio  
71              0            0  
72              0            0  
73              0            0  
74              0            0  
75              0            0  

[5 rows x 255 columns]

Group the dataframe by the "Nborhood" column using the mean of each "venue category":

In [177]:
toronto_grouped = toronto_onehot.groupby('Nborhood').mean().reset_index()
toronto_grouped.head()

Nborhood  Accessories Store  Adult Boutique  Afghan Restaurant  Airport  \
0      M1W                0.0             0.0                0.0      0.0   
1      M2J                0.0             0.0                0.0      0.0   
2      M2N                0.0             0.0                0.0      0.0   
3      M3C                0.0             0.0                0.0      0.0   
4      M3H                0.0             0.0                0.0      0.0   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                 0.0           0.0             0.0              0.0   
1                 0.0           0.0             0.0              0.0   
2                 0.0           0.0             0.0              0.0   
3                 0.0           0.0             0.0              0.0   
4                 0.0           0.0             0.0              0.0   

   Airport Terminal     ...       Train Station  \
0               0.0     ...                 0.0   
1               0.0     ...                 0.0   
2               0.0     ...                 0.0   
3               0.0     ...                 0.0   
4               0.0     ...                 0.0   

   Vegetarian / Vegan Restaurant  Video Game Store  Video Store  \
0                            0.0          0.000000     0.000000   
1                            0.0          0.016129     0.000000   
2                            0.0          0.000000     0.000000   
3                            0.0          0.000000     0.000000   
4                            0.0          0.000000     0.058824   

   Vietnamese Restaurant  Warehouse Store  Wine Bar  Wings Joint  \
0               0.000000              0.0       0.0     0.000000   
1               0.000000              0.0       0.0     0.016129   
2               0.028571              0.0       0.0     0.000000   
3               0.000000              0.0       0.0     0.000000   
4               0.000000              0.0       0.0     0.000000   

   Women's Store  Yoga Studio  
0       0.000000          0.0  
1       0.016129          0.0  
2       0.000000          0.0  
3       0.000000          0.0  
4       0.000000          0.0  

[5 rows x 255 columns]

Function to return the most frequent categories en each row of the dataframe:

In [178]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Build a new dataframe with the top 10 categories for each neighborhood:

In [179]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Nborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
0          M1W  Fast Food Restaurant    Chinese Restaurant   
1          M2J        Clothing Store  Fast Food Restaurant   
2          M2N           Coffee Shop            Restaurant   
3          M3C                   Gym      Asian Restaurant   
4          M3H           Coffee Shop              Pharmacy   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0           Pizza Place   American Restaurant        Breakfast Spot   
1           Coffee Shop            Shoe Store            Food Court   
2      Ramen Restaurant        Sandwich Place         Shopping Mall   
3           Coffee Shop            Beer Store        Clothing Store   
4    Frozen Yogurt Shop  Fast Food Restaurant        Sandwich Place   

  6th Most Common Venue 7th Most Common Venue  8th Most Common Venue  \
0        Sandwich Place     Electronics Store             Nail Salon   
1      Toy / Game Store                Bakery             Restaurant   
2                  Café      Sushi Restaurant  Indonesian Restaurant   
3    Chinese Restaurant    Dim Sum Restaurant             Restaurant   
4                 Diner           Bridal Shop                   Bank   

  9th Most Common Venue 10th Most Common Venue  
0          Camera Store                    Gym  
1   Japanese Restaurant          Burrito Place  
2                Lounge         Ice Cream Shop  
3              Bus Line         Discount Store  
4            Restaurant            Supermarket

Apply KMEANS to cluster the toronto_grouped dataframe. We have to delete the "Nborhood" column because KMEANS only accepts numerical data. 

We tried with several number of clusters and finally decided to use 6 clusters.

In [180]:
# set number of clusters
kclusters = 6

toronto_grouped_clustering = toronto_grouped.drop('Nborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([1, 3, 3, 1, 1, 1, 3, 1, 3, 1], dtype=int32)

Add cluster number to neighborhoods_venues_sorted dataframe and merge it with the latlon_df dataframe which was created in "part 2" of this notebook.

We use "inner" join since we filtered the neighborhoods to keep only those with more than 10 venues.

In [181]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = latlon_df

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.set_index('Postcode').join(neighborhoods_venues_sorted.set_index('Neighborhood'), how="inner")

toronto_merged.reset_index(inplace=True)
toronto_merged.head() 

index      Borough                                    Neighbourhood  \
0   M1W  Scarborough                                  L'Amoreaux West   
1   M2J   North York                     Oriole, Henry Farm, Fairview   
2   M2N   North York                                 Willowdale South   
3   M3C   North York                 Don Mills South, Flemingdon Park   
4   M3H   North York  Downsview North, Bathurst Manor, Wilson Heights   

    Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0  43.799525 -79.318389               1  Fast Food Restaurant   
1  43.778517 -79.346556               3        Clothing Store   
2  43.770120 -79.408493               3           Coffee Shop   
3  43.725900 -79.340923               1                   Gym   
4  43.754328 -79.442259               1           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0    Chinese Restaurant           Pizza Place   American Restaurant   
1  Fast Food Restaurant           Coffee Shop            Shoe Store   
2            Restaurant      Ramen Restaurant        Sandwich Place   
3      Asian Restaurant           Coffee Shop            Beer Store   
4              Pharmacy    Frozen Yogurt Shop  Fast Food Restaurant   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0        Breakfast Spot        Sandwich Place     Electronics Store   
1            Food Court      Toy / Game Store                Bakery   
2         Shopping Mall                  Café      Sushi Restaurant   
3        Clothing Store    Chinese Restaurant    Dim Sum Restaurant   
4        Sandwich Place                 Diner           Bridal Shop   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0             Nail Salon          Camera Store                    Gym  
1             Restaurant   Japanese Restaurant          Burrito Place  
2  Indonesian Restaurant                Lounge         Ice Cream Shop  
3             Restaurant              Bus Line         Discount Store  
4                   Bank            Restaurant            Supermarket

Finally, we display a map with the neighborhoods colored by the cluster to which they belong.

In [182]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['index'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters